src/gc-heap-snapshot.cpp

```diff
@@ -0,0 +1,420 @@
+//}
+
+HeapSnapshot *g_snapshot = nullptr;
+
+JL_DLLEXPORT int count_nodes = 0;
+JL_DLLEXPORT int count_edges = 0;
+
+JL_DLLEXPORT void jl_gc_take_heap_snapshot(JL_STREAM *stream) {
+    // Enable snapshotting
+    HeapSnapshot snapshot;
+    g_snapshot = &snapshot;
+    gc_heap_snapshot_enabled = true;
+
+    _add_internal_root(&snapshot);
+
+    // Do GC, which will callback into record_edge_to_gc_snapshot()...
+    jl_gc_collect(JL_GC_FULL);
+
+    // Disable snapshotting
+    gc_heap_snapshot_enabled = false;
+    g_snapshot = nullptr;
+
+    // When we return, the snapshot is full
+    // Dump the snapshot
+    serialize_heap_snapshot(stream, snapshot);
+
+    // Debugging
+    //jl_printf(JL_STDERR, "nodes: %d\n", count_nodes);
+    //jl_printf(JL_STDERR, "edges: %d\n", count_edges);
+}
+
+// adds a node at id 0 which is the "uber root":
+// a synthetic node which points to all the GC roots.
+void _add_internal_root(HeapSnapshot *snapshot) {
+    Node internal_root{
+        snapshot->node_types.find_or_create_string_id("synthetic"),
+        "(internal root)", // name
+        0, // id
+        1, // size
+
+        0, // size_t trace_node_id (unused)
+        0, // int detachedness;  // 0 - unknown,  1 - attached;  2 - detached
+
+        // outgoing edges
+        vector<Edge>(),
+    };
+    snapshot->nodes.push_back(internal_root);
+}
+
+// mimicking https://github.com/nodejs/node/blob/5fd7a72e1c4fbaf37d3723c4c81dce35c149dc84/deps/v8/src/profiler/heap-snapshot-generator.cc#L597-L597
+void record_node_to_gc_snapshot(jl_value_t *a) JL_NOTSAFEPOINT {
+    auto val = g_snapshot->node_ptr_to_index_map.find((void*)a);
+    if (val != g_snapshot->node_ptr_to_index_map.end()) {
+        return;
+        //return &g_snapshot->nodes[val->second];
+    }
+
+    // Insert a new Node
+    size_t self_size = 0;
+    string name = "<missing>";
+    string node_type = "object";
+
+    if (a == (jl_value_t*)jl_malloc_tag) {
+        name = "<malloc>";
+    } else {
+        jl_datatype_t* type = (jl_datatype_t*)jl_typeof(a);
+
+        if ((uintptr_t)type < 4096U) {
+            name = "<unkown>";
+        } else if (type == (jl_datatype_t*)jl_buff_tag) {
+            name = "<buffer>";
+        } else if (type == (jl_datatype_t*)jl_malloc_tag) {
+            name = "<malloc>";
+        } else if (jl_is_string(a)) {
+            node_type = "string";
+            name = jl_string_data(a);
+            self_size = jl_string_len(a);
+        } else if (jl_is_symbol(a)) {
+            node_type = "symbol";
+            name = jl_symbol_name((jl_sym_t*)a);
+            self_size = name.length();
+        } else if (jl_is_datatype(type)) {
+            self_size = (size_t)jl_datatype_size(type);
+            // TODO: get the entire type, including type parameters
+            // - option 1: jl_static_show it here:
+            //   crashes because it might do GC, and we are already inside GC.
```

Comment:
! that's good to know! @vchuravy and @vilterp maybe we can turn this back on then.

@vtjnash we were also concerned about performance, thinking that it might be better to unfreeze the world as soon as we can.. We were thinking about maybe actually trying to create a GC reference to all the values we want to print, so that we can safely continue accessing them after GC, without worrying about them being released by another GC phase in the meantime, and then printing them and then releasing our references.

But A) that seems complicated, B) i guess it's probably not kosher to add GC structure _during_ the mark phase?, so we left it out for now.

And now that you're saying it should be safe to do inside GC, maybe we just go back to doing the simple thing and do the printing inside there 👍

Reply:
We were seeing crashes and hangs, i think, when this wasn't commented out.. Maybe they were actually just symptoms of a memory violation somewhere else though... 🤔 something to look into! Or maybe we were using the wrong API for printing?